### PR TITLE
Add boundary condition support to Rectangle domain creator

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Add boundary condition support to Rectangle domain creator

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Domain/Creators/Rectangle.hpp [hunk 1]
```diff
@@ -75,21 +77,50 @@ class Rectangle : public DomainCreator<2> {
     static constexpr Options::String help = {
         "The time dependence of the moving mesh domain."};
   };
+  template <typename BoundaryConditionsBase>
+  struct BoundaryCondition {
+    static std::string name() noexcept { return "BoundaryCondition"; }
+    static constexpr Options::String help =
+        "Options for the boundary conditions.";
```
Comment (nilsvu) [inline src/Domain/Creators/Rectangle.hpp]: Add something like "... to impose on all sides"

File: src/Domain/Creators/Rectangle.hpp [hunk 2]
```diff
@@ -75,21 +77,50 @@ class Rectangle : public DomainCreator<2> {
     static constexpr Options::String help = {
         "The time dependence of the moving mesh domain."};
   };
+  template <typename BoundaryConditionsBase>
+  struct BoundaryCondition {
```
Comment (nilsvu) [inline src/Domain/Creators/Rectangle.hpp:81]: Do I understand correctly that you're adding basic boundary condition support to impose the same conditions on all sides, and then we can generalize this later to treat each side differently? (I think that's a good strategy 👍)

File: tests/Unit/Domain/Creators/Test_Rectangle.cpp [hunk 3]
```diff
@@ -142,53 +175,117 @@ void test_rectangle() {
            {Direction<2>::lower_eta(), {0, aligned_orientation}},
            {Direction<2>::upper_eta(), {0, aligned_orientation}}}},
       std::vector<std::unordered_set<Direction<2>>>{{}});
+
+  const creators::Rectangle periodic_rectangle_boundary_conditions{
```
Comment (nilsvu) [inline tests/Unit/Domain/Creators/Test_Rectangle.cpp]: (optional) I think these tests would benefit from a few added scopes like `{ INFO("Periodic"); ...}`, so they print a little context when they fail and so variable names don't have to be prefixed with e.g. `periodic`. I'm happy to merge this as-is, and suggest to add some more scopes in the upcoming domain creator PRs.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Domain/Creators/Rectangle.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Domain/Creators/Rectangle.hpp:81]: Yep, that's right! There are just way too many possible combinations, especially for domain creators like 3d aligned lattice, that I think it's better to just wait until someone needs a specific config to implement.

**Comment by nilsdeppe** [comment]
Rebased on develop and pushed a fixup. Thanks for the review :)

**Review by nilsvu — DISMISSED** [review]
LGTM, please squash

**Review by kidder — COMMENTED** [review]
LGTM

**Comment by nilsdeppe** [comment]
Rebased on develop and squashed. Thanks for the reviews :)

**Review by kidder — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]

**Pull request merged**